Modules/booking/Controller/BookingsupsabstractController.php

```diff
@@ -0,0 +1,186 @@
+        $supID = is_array($supID) ? $supID : [$supID];
+        $supResources = is_array($supResources) ? $supResources : [$supResources];
+
+        if ($this->invoicable && $this->hasInvoicingUnitsDuplicates($supResources, $supIsInvoicingUnit, $id_space, $lang)) {
+            // find out if multiple sups are used as invoicing units
+            return $this->redirect("booking" . $this->formUrl ."/".$id_space);
+        }
+
+        $supacks = [];
+        $id_sups = [];
+        $coupleSupResourceExists = false;
+        for ($i = 0; $i < count($supID); $i++) {
+            if($supName[$i] == "") {
+                continue;
+            } else if ($supID[$i]) {
+                $supacks[$supName[$i]] = $supID[$i];
+            }
+            if (!$supID[$i]) {
+                // If sup id not set, use from known sups
+                if(isset($supacks[$supName[$i]])) {
+                    $supID[$i] = $supacks[$supName[$i]];
+                    if ($this->coupleSupResourceExists($supID[$i],$supResources[$i], $id_space)) {
+                        $coupleSupResourceExists = [
+                            "resource" => $modelResource->get($id_space, $supResources[$i])['name'],
+                            "sup" => $supName[$i]
+                        ];
+                    }
+                } else {
+                    // Or create a new sup
+                    $cvm = new CoreVirtual();
+                    $vid = $cvm->new($this->supsType);
+                    $supID[$i] = $vid;
+                    $supacks[$supName[$i]] = $vid;
+                }
+            }
+
+            $this->modelSups->setSupplementary($id_space, $supID[$i], $supResources[$i], $supName[$i], $supMandatory[$i] ?? 0, $supIsInvoicingUnit[$i] ?? 0, $supDuration[$i] ?? 0);
+            array_push($id_sups, $this->modelSups->getBySupID($id_space, $supID[$i], $supResources[$i])['id']);
+        }
+        
+        // If package in db is not listed in provided package list, delete them
+        $this->modelSups->removeUnlisted($id_space, $id_sups, false);
+        $this->handleMessages($coupleSupResourceExists, $lang);
+    }
+
+    protected function coupleSupResourceExists($id_sup, $id_resource, $id_space) {
+        $dbSups = $this->modelSups->getByResource($id_space, $id_resource);
+        foreach ($dbSups as $dbSup) {
+            if ($dbSup['id_' . $this->supsType] == $id_sup) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    protected function hasInvoicingUnitsDuplicates($supResources, $supIsInvoicingUnit, $id_space, $lang) {
+        $result = false;
+        $invoicingUnitsResources = [];
+        foreach ($supResources as $index => $resource) {
+            if ($supIsInvoicingUnit[$index] == 1) {
+                if (in_array($resource, $invoicingUnitsResources)) {
+                    $_SESSION["flash"] = BookingTranslator::maxInvoicingUnits($lang);
```

Comment:
is message adapted to package/supp/....?

Reply:
$invoicable property is set to false for supplementaries and packages. So it never pass through hasInvoicingUnitsDuplicates() (and these lines of code). cf line 101:
```
if ($this->invoicable && $this->hasInvoicingUnitsDuplicates($supResources, $supIsInvoicingUnit, $id_space, $lang))
```